README.md

```diff
@@ -95,6 +95,75 @@ $ sudo cat /sys/kernel/debug/tracing/trace_pipe
            <...>-461101 [018] d... 505434.345367: bpf_trace_printk: UPROBE EXIT: return = 5
 ```
 
+# Kprobe
+
+`kprobe` is an example of dealing with kernel-space entry and exit (return) probes,
+`kprobe` and `kretprobe` in libbpf lingo. It attaches `kprobe` and `kretprobe`
+BPF programs to the `bprm_execve` function and logs the PID, filename
+and return result, respectively, using `bpf_printk()` macro.
+
+```shell
+$ sudo ./kprobe
+libbpf: loading object 'kprobe_bpf' from buffer
+...
+Successfully started!
+...........
+```
+
+The `kprobe` demo output in `/sys/kernel/debug/tracing/trace_pipe` should look something like this:
+
+```shell
+$ sudo cat /sys/kernel/debug/tracing/trace_pipe
+             cat-3823    [001]   4047.478649: bpf_trace_printk: KPROBE ENTRY pid = 3823, filename = /usr/bin/cat
+
+             cat-3823    [001]   4047.479141: bpf_trace_printk: KPROBE EXIT: ret = 0
+
+              i3-3825    [004]   4048.568096: bpf_trace_printk: KPROBE ENTRY pid = 3825, filename = /bin/sh
+
+              sh-3825    [004]   4048.569349: bpf_trace_printk: KPROBE EXIT: ret = 0
+
+              sh-3825    [004]   4048.574687: bpf_trace_printk: KPROBE ENTRY pid = 3825, filename = /usr/bin/i3-sensible-terminal
+```
+
+# Fentry
+`fentry` is an example that uses fentry and fexit tracing. It attaches `fentry` and `fexit` traces to `bprm_execve`
+which is called by the execve family of functions and logs the return value, PID and filename to the trace pipe.
+Important differences compared to kprobes are improved performance and usability. fentry and fexit programs are available
+starting from 5.5 kernels.
+In this example easier usability is shown with the ability to directly dereference pointer arguments like in normal C,
+instead of using various read helpers.
+The second difference is the requirement to list all the function arguments, followed by the return value,
```

Comment:
it's not really a requirement, you can do just:

int my_fentry(void *ctx) { ... }

and it will work just fine. So would be good to mild it down this sentence a bit